keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+For pruning and defaulting we have to apply the OpenAPI validation schema inside of the decoder step (the left box inside apiextensions-apiserver in the figure above). This is considerably earlier than for native types. This has a number of implications:
+* We have to apply full OpenAPI value validation (e.g. regular expressions, propositional evaluation) during decoding.
+* Our generic registry applies validation after defaulting. We need the other way around for CRDs.
+* Our generic registry expects defaulting to always succeed (no error result type). CRD validation needed by defaulting can fail.
+
+To avoid these and to get sane semantics, **we propose to split the CRD validation into two steps:**
+1. During decoding we validate using a skeleton schema which lacks value validations and propositional logic (`anyOf`, `allOf`, `oneOf`, `not`).
+2. During defaulting we validate using the skeleton schema and apply the resulting defaulter.
+During standard generic registry validation phase we validate using the full validation schema.
+
+Because the first step succeeds, also the second step succeeds. So we don’t need error handling during defaulting. The third step will catch wrong types of the used defaults.
+
+## Formal Proposal – following pruning option 1
+
+The examples in the motivational section show that the semantics of full OpenAPI validation schemata are not trivial in respect to pruning and defaulting. To simplify the algorithms and to enforce "sane" schemata which allow to split schema and value validation, we propose to derive a skeleton schema from the full user-given OpenAPI validation schema
+* which does not contain value validations
+* which is complete enough for pruning and defaulting.
+
+**Remark:** we have two options to define and implement pruning and defaulting:
+1. directly on the full OpenAPI validation schema with two custom algorithms doing parallel recursion over the schema and the input object.
+2. via an intermediate representation (the skeleton schema) and using go-openapi pruning and defaulting. Both algorithms are ten-liners based on the go-openapi/validate output. With the skeleton schema the go-openapi pruning algorithm coincides with our pruning option 1.
+
+Both routes lead to the same algorithm: the intermediate representation of the skeleton schema makes merging of OpenAPI validation schema constraints explicit, while the custom algorithms would hide that in its recursion code.
+
+Moreover, the main reason for the intermediate schema: the custom algorithm would have to replicate a lot of the validation logic of go-openapi, e.g. the semantics of `properties`, `additionalPropoerties`, `patternProperties` and the same for items of an array. With route 2 we get all of this for free. 
```

Comment:
Looks like we reject `patternProperties` today completely, although it exists in JSONSchemaProps. So for completeness it does not hurt to keep it in the doc here, even though in practice it does not matter. Maybe some day we add patternProperties.